### ISSUE TITLE
mooseError is not flushing ConsoleStream in OutputWarehouse

### ISSUE THREAD

**Issue by YaqiWang** [idaholab/moose#6615]
### Description of the enhancement or error report

Currently calling `mooseError()` will cause program to be terminated without flushing the buffer of `_console`, which could potentially lose very important information on screen.
### Rationale for the enhancement or information for reproducing the error

This can be easily seen with two lines at the beginning of `Steady::execute()` and run any test using `Steady`.

```
  _console << "some extremely important information\n";
  mooseError("stop! did we see the above information?");
```

Although the message can be flushed explicitly, there could be cases that doing an explicitly flush is too demanding or even impossible (users are directly accessing the buffer through `OutputWarehouse::consoleBuffer()` for example).

In summary, somebody could output bunch of information in `_console` without flushing, then these information would be lost once `mooseError` is called.
### Identified impact

(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
This seems to me a design flaw and can result codes with lesser maintainability.

**Comment by YaqiWang** [comment]
A possibly related minor issue about `mooseWarning` is that the message in the warning is not properly prefixed with the subapp name.

**Comment by permcody** [comment]
Just thinking out loud here: I know that there is a flush in the dtor of the console stream. Perhaps this isn't getting called under certain `MPI_Abort()` conditions. It may be worth trying an explicit flush of the console stream in the `mooseError()` macro.

**Comment by YaqiWang** [comment]
I am not sure if you are aware of `OStreamProxy` in libMesh. That can make `_console` an `ostream`. I was also thinking can we duplicate `mooseError` and `mooseWarning` in `ConsoleStreamInterface`? But I think this is out of my hand to have a proper fix for these. So I will rely on you for the fix. Thanks!

**Comment by rwcarlsen** [comment]
Fixed by #8411

**Comment by YaqiWang** [comment]
@rwcarlsen @permcody I think this is only partly fixed. If you do the same thing in `Steady` constructor, you will see that the error message before mooseError will be swallowed.  It looks like if the error happens before a certain point (I guess the construction of `Console` which is quite late in the actions), issue still exists.

**Comment by YaqiWang** [comment]
BTW, I do not know why I cannot reopen issues created by me.

**Comment by YaqiWang** [comment]
I am speechless. Do we need to fight this annoying output issues this hard?

**Comment by permcody** [comment]
Yes - I/O is extremely expensive (1-2 orders of magnitude) more expensive then normal binary ops, possibly more. It's worth buffering things and formatting things the way we want them. With subapps it's very nice to be able to indent to make the output look nice so it makes all of this very difficult to get right.

The system works fine most of the time (as is evident by the date on this last fix), but there are always going to be a few edges cases when you abort really early on. We'll just have to figure these out one at a time.
